File: backend/app/schemas/roast.py
from pydantic import BaseModel, Field


class RoastRequest(BaseModel):
    repo_url: str = Field(
        ..., pattern=r"^https://github\.com/[a-zA-Z0-9\-_.]+/[a-zA-Z0-9\-_.]+/?$"
    )
    brutality_level: int = Field(..., ge=1, le=5)


class RoastSubmitResponse(BaseModel):
    id: str
    status: str
    repo_url: str
    repo_owner: str
    repo_name: str
    brutality_level: int
    created_at: str


class RepoMetadata(BaseModel):
    stars: int
    forks: int
    language: str | None
    size_kb: int
    open_issues: int
    description: str | None
    topics: list[str]
    default_branch: str
    last_push: str
    has_wiki: bool
    license: str | None


class AnalysisFinding(BaseModel):
    category: str
    severity: str
    finding: str
    evidence: str


class RoastCategory(BaseModel):
    name: str
    score: int
    emoji: str
    roast: str
    suggestions: list[str]


class RoastResult(BaseModel):
    overall_score: int
    letter_grade: str
    summary: str
    top_burns: list[str]
    categories: list[RoastCategory]


class RoastResponse(BaseModel):
    id: str
    status: str
    repo_url: str
    repo_owner: str
    repo_name: str
    brutality_level: int
    error_message: str | None = None
    repo_metadata: RepoMetadata | None = None
    roast_result: RoastResult | None = None
    overall_score: int | None = None
    letter_grade: str | None = None
    created_at: str
    completed_at: str | None = None


class RoastFeedItem(BaseModel):
    id: str
    repo_url: str
    repo_owner: str
    repo_name: str
    brutality_level: int
    overall_score: int
    letter_grade: str
    top_burns: list[str]
    repo_metadata: RepoMetadata
    completed_at: str


class RoastFeedResponse(BaseModel):
    roasts: list[RoastFeedItem]
    total: int
    limit: int
    offset: int

File: frontend/src/types/roast.ts
export interface RepoMetadata {
  stars: number;
  forks: number;
  language: string | null;
  size_kb: number;
  open_issues: number;
  description: string | null;
  topics: string[];
  default_branch: string;
  last_push: string;
  has_wiki: boolean;
  license: string | null;
}

export interface RoastCategory {
  name: string;
  score: number;
  emoji: string;
  roast: string;
  suggestions: string[];
}

export interface RoastResult {
  overall_score: number;
  letter_grade: string;
  summary: string;
  top_burns: string[];
  categories: RoastCategory[];
}

export interface RoastResponse {
  id: string;
  status: "pending" | "analyzing" | "roasting" | "complete" | "failed";
  repo_url: string;
  repo_owner: string;
  repo_name: string;
  brutality_level: number;
  error_message: string | null;
  repo_metadata: RepoMetadata | null;
  roast_result: RoastResult | null;
  overall_score: number | null;
  letter_grade: string | null;
  created_at: string;
  completed_at: string | null;
}

export interface RoastSubmitResponse {
  id: string;
  status: string;
  repo_url: string;
  repo_owner: string;
  repo_name: string;
  brutality_level: number;
  created_at: string;
}

export interface RoastFeedItem {
  id: string;
  repo_url: string;
  repo_owner: string;
  repo_name: string;
  brutality_level: number;
  overall_score: number;
  letter_grade: string;
  top_burns: string[];
  repo_metadata: RepoMetadata;
  completed_at: string;
}

export interface RoastFeedResponse {
  roasts: RoastFeedItem[];
  total: number;
  limit: number;
  offset: number;
}
